lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py

```diff
@@ -319,45 +319,89 @@ def ensure_tags(conn, module, subnet, tags, purge_tags):
+        if module.params['wait']:
+            updated = False
+            while module.params['wait_timeout'] + start_time - time.time() > 0:
+                current_value = conn.describe_subnets(SubnetIds=[subnet['id']])['Subnets'][0]['MapPublicIpOnLaunch']
+                if current_value != map_public:
+                    time.sleep(3)
+                else:
+                    updated = True
+                    break
+            if not updated:
+                module.warn("Failed to wait for MapPublicIpOnLaunch to be updated for {0}".format(subnet['id']))
     except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
         module.fail_json_aws(e, msg="Couldn't modify subnet attribute")
 
 
-def ensure_assign_ipv6_on_create(conn, module, subnet, assign_instances_ipv6, check_mode):
+def ensure_assign_ipv6_on_create(conn, module, subnet, assign_instances_ipv6, check_mode, start_time):
     if check_mode:
         return
-
     try:
         conn.modify_subnet_attribute(SubnetId=subnet['id'], AssignIpv6AddressOnCreation={'Value': assign_instances_ipv6})
+        if module.params['wait']:
+            updated = False
+            while module.params['wait_timeout'] + start_time - time.time() > 0:
+                current_value = conn.describe_subnets(SubnetIds=[subnet['id']])['Subnets'][0]['AssignIpv6AddressOnCreation']
+                if current_value != assign_instances_ipv6:
+                    time.sleep(3)
+                else:
+                    updated = True
+                    break
+            if not updated:
+                module.warn("Failed to wait for AssignIpv6AddressOnCreation to be updated for {0}".format(subnet['id']))
     except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
         module.fail_json_aws(e, msg="Couldn't modify subnet attribute")
 
 
-def disassociate_ipv6_cidr(conn, module, subnet):
+def disassociate_ipv6_cidr(conn, module, subnet, start_time):
     if subnet.get('assign_ipv6_address_on_creation'):
-        ensure_assign_ipv6_on_create(conn, module, subnet, False, False)
+        ensure_assign_ipv6_on_create(conn, module, subnet, False, False, start_time)
 
     try:
         conn.disassociate_subnet_cidr_block(AssociationId=subnet['ipv6_association_id'])
     except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
         module.fail_json_aws(e, msg="Couldn't disassociate ipv6 cidr block id {0} from subnet {1}"
                              .format(subnet['ipv6_association_id'], subnet['id']))
 
+    # Wait for cidr block to be disassociated
+    if module.params['wait']:
+        remaining_wait_timeout = module.params['wait_timeout'] + start_time - time.time()
+        waiter_config = {'Delay': 5, 'MaxAttempts': remaining_wait_timeout // 5}
+        try:
+            conn.get_waiter('subnet_available').wait(Filters=ansible_dict_to_boto3_filter_list(
```

Comment:
Looking for a way to optimize this - watching the AWS console while the tests run, this seems to take much longer than it should.